packages/asset-server-plugin/src/s3-asset-storage-strategy.ts

```diff
@@ -157,113 +157,114 @@ export function configureS3AssetStorage(s3Config: S3Config) {
 
         const config = {
-            credentials: this.getS3Credentials(),
+            credentials: await this.getS3Credentials(),
             ...this.s3Config.nativeS3Configuration,
-        };
+        } satisfies S3ClientConfig
+
         this.s3 = new this.AWS.S3(config);
         await this.ensureBucket(this.s3Config.bucket);
     }
 
     destroy?: (() => void | Promise<void>) | undefined;
 
     async writeFileFromBuffer(fileName: string, data: Buffer): Promise<string> {
-        const result = await this.s3
-            .upload(
-                {
-                    Bucket: this.s3Config.bucket,
-                    Key: fileName,
-                    Body: data,
-                },
-                this.s3Config.nativeS3UploadConfiguration,
-            )
-            .promise();
-        return result.Key;
+        const upload = new (await import('@aws-sdk/lib-storage')).Upload({
+            client: this.s3,
+            params: {
+                ...this.s3Config.nativeS3UploadConfiguration,
+                Bucket: this.s3Config.bucket,
+                Key: fileName,
+                Body: data,
+            }
+        })
+
+        return upload.done().then((result) => {
+            if (!('Key' in result) || !result.Key) {
+                Logger.error(`Got undefined Key for ${fileName}`, loggerCtx);
+                throw new Error(`Got undefined Key for ${fileName}`);
+            }
+
+            return result.Key;
+        })
     }
 
-    async writeFileFromStream(fileName: string, data: Stream): Promise<string> {
-        const result = await this.s3
-            .upload(
-                {
-                    Bucket: this.s3Config.bucket,
-                    Key: fileName,
-                    Body: data,
-                },
-                this.s3Config.nativeS3UploadConfiguration,
-            )
-            .promise();
-        return result.Key;
+    async writeFileFromStream(fileName: string, data: Readable): Promise<string> {
+        const upload = new (await import('@aws-sdk/lib-storage')).Upload({
+            client: this.s3,
+            params: {
+                ...this.s3Config.nativeS3UploadConfiguration,
+                Bucket: this.s3Config.bucket,
+                Key: fileName,
+                Body: data,
+            },
+        });
+
+        return upload.done().then((result) => {
+            if (!('Key' in result) || !result.Key) {
+                Logger.error(`Got undefined Key for ${fileName}`, loggerCtx);
+                throw new Error(`Got undefined Key for ${fileName}`);
+            }
+
+            return result.Key;
+        });
     }
 
     async readFileToBuffer(identifier: string): Promise<Buffer> {
-        const result = await this.s3.getObject(this.getObjectParams(identifier)).promise();
-        const body = result.Body;
+        const result = await this.s3.getObject(this.getObjectParams(identifier));
+        const body = result.Body as Readable | undefined;
```

Comment:
Why is the type assertion needed here? From [the s3 client docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/getobjectcommandoutput.html#body) I see it returns type `SdkStream<undefined | Readable | Blob | ReadableStream<any>>` Is no special handling needed for Blob?

Reply:
Because by default it also returns the types for the browser API types which are Blob and ReadableStream

Reply:
Ah ok, makes sense 👍